v2/protocol/http/protocol.go

```diff
@@ -133,6 +145,64 @@ func (p *Protocol) Request(ctx context.Context, m binding.Message) (binding.Mess
 	return NewMessage(resp.Header, resp.Body), NewResult(resp.StatusCode, "%w", result)
 }
 
+func (p *Protocol) doWithRetry(backoff *types.Backoff) func(*http.Request) (binding.Message, error) {
+	return func(req *http.Request) (binding.Message, error) {
+		retries := 0
+		var results []protocol.Result
+		for {
+			resp, err := p.Client.Do(req)
+
+			// Fast track common case.
+			if err == nil && resp.StatusCode/100 == 2 {
+				result := NewResult(resp.StatusCode, "%w", protocol.ResultACK)
+
+				if retries != 0 {
+					result = NewRetriesResult(result, retries, results)
+				}
+				return NewMessage(resp.Header, resp.Body), result
+			}
+
+			// Slow case.
+			retry := retries < backoff.Retry
+
+			var result protocol.Result
+			if err != nil {
+				result = protocol.NewReceipt(false, "%w", err)
+				if retry && !err.(*url.Error).Timeout() {
+					// Do not retry if the error is not a timeout
+					retry = false
+				}
+			} else {
+				// No error, status code is not 2xx
+				result = NewResult(resp.StatusCode, "%w", protocol.ResultNACK)
+
+				// Potentially retry when:
+				// - 413 Payload Too Large with Retry-After (NOT SUPPORTED)
+				// - 425 Too Early
+				// - 429 Too Many Requests
+				// - 503 Service Unavailable (with or without Retry-After) (IGNORE Retry-After)
+				// - 504 Gateway Timeout
+
+				sc := resp.StatusCode
+				if retry && sc != 425 && sc != 429 && sc != 503 && sc != 504 {
+					// Permanent error
+					retry = false
+				}
+			}
+
+			if !retry {
+				return NewMessage(resp.Header, resp.Body), NewRetriesResult(result, retries, results)
+			}
+
+			results = append(results, result)
+			retries++
+
+			// Linear backoff.
+			time.Sleep(backoff.Delay)
```

Comment:
I would make this method more of a wrapper for doOnce with a some kind of pattern of:


```go
for { 
  select {
    case <-ctx.Done():
    // canceled.
    case <-timer.Tick():
      strongTypedErr: doOnce(ctx, request)
      ... custom logic to see if we should stop retry
  }
}
```

That way the logic of producing the `protocol.NewReceipt` is only one place, and you are still free to collect them in the outer retry method.

Reply:
I tried but the reason I didn't call `doOnce` is because of `if retry && !err.(*url.Error).Timeout() {`. When `doOnce` returns an error, there is no way to unwrap the original error.

Reply:
which can be easily solved by changing `doOnce` signature.

Reply:
ended up trying go 1.13 unwrap API. Not tested yet.